### ISSUE TITLE
<CGAL/Arr_polycurve_basic_traits_2.h> use deprecated classes

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#222]
https://cgal.geometryfactory.com/CGAL/Members/testsuite/CGAL-4.7-Ic-102/Arrangement_on_surface_2/TestReport_lrineau_Fedora-strict-ansi.gz

```
In file included from /mnt/testsuite/include/CGAL/Arr_polycurve_basic_traits_2.h:35:0,
                 from /mnt/testsuite/include/CGAL/Arr_polycurve_traits_2.h:35,
                 from /mnt/testsuite/include/CGAL/Arr_polyline_traits_2.h:37,
                 from /home/cgal_tester/build/src/cmake/platforms/Fedora-strict-ansi/test/Arrangement_on_surface_2/test_geom_traits.h:16,
                 from /home/cgal_tester/build/src/cmake/platforms/Fedora-strict-ansi/test/Arrangement_on_surface_2/test_traits.cpp:64:
/mnt/testsuite/include/CGAL/Arr_geometry_traits/Polycurve_2.h:265:32: warning: 'const_iterator' is deprecated [-Wdeprecated-declarations]
     const_iterator& operator++()
                                ^
/mnt/testsuite/include/CGAL/Arr_geometry_traits/Polycurve_2.h:205:25: note: declared here
   class CGAL_DEPRECATED const_iterator {
                         ^
/mnt/testsuite/include/CGAL/Arr_geometry_traits/Polycurve_2.h:271:34: warning: 'const_iterator' is deprecated [-Wdeprecated-declarations]
     const_iterator operator++(int)
                                  ^
```

**Comment by lrineau** [comment]
If the test program tests the deprecated class on purpose, then it should define `CGAL_NO_DEPRECATION_WARNINGS`.

**Comment by efifogel** [comment]
It is on purpose.
I'll add the necessary code.
Thanks.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
I am worried that that warning has been seen in:
```
Arrangement_on_surface_2
Arrangement_on_surface_2_Demo
Arrangement_on_surface_2_Examples
CGAL_ipelets_Demo
Snap_rounding_2
Snap_rounding_2_Demo
Snap_rounding_2_Examples
Sweep_line_2
```
See the column "ArchLinux" in [CGAL-4.7-Ic-120 test results](https://cgal.geometryfactory.com/CGAL/Members/testsuite/results-4.7-Ic-120.shtml) for example.

**Comment by efifogel** [comment]
ok, I'm not sure exactly what this warning is about.
It comes from the compiler. It merely indicates that deprecated code is
compiled in.
Indeed, there is deprecated code.
However, the test program never execute the deprecated code.
(BW, it seems like gcc 5.1.1, right?)

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
I can partially answer to you @efifogel. The machine is a recent Linux Fedora, with g++-5.1.1, and the compilation flags are `-Wall -Wextra -ansi -frounding-math`.

**Comment by efifogel** [comment]
I think the solution is to suppress this warning, because we do have
deprecated code (and we want to have deprecated code); see
"-Wdeprecated-declarations"

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
I disagree. Unless there is a bug in the compiler, the warning is suppose to warn about deprecated code that is *used*. I will try to understand more about the scenario. I cannot do that today, unfortunately.

**Comment by efifogel** [comment]
By 'used' you mean compiled in.
If deprecated code is completely generic, then it should not be compiled
in, and as you say, the warning is legitimate.
However, if it's not generic, then the deprecated code is still linked and
in this case, the warning is still correct, but irrelevant.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
It that case, the deprecated code is composed of template classes, or methods of template classes. It should not be compiled if not used.

What surprises me if the lack of context of the warnings.

Let's forget that bug for a while, until I can find time to reproduce it and dig a bit.

**Comment by efifogel** [comment]
On Mon, Aug 24, 2015 at 12:29 PM, Laurent Rineau <notifications@github.com>
wrote:

> It that case, the deprecated code is composed of template classes, or
> methods of template classes. It should not be compiled if not used.
>
> What surprises me if the lack of context of the warnings.
>
> Let's forget that bug for a while, until I can find time to reproduce it
> and dig a bit.
>
Sure. Actually, I was about to suggest it, at least until I get a hold of
g++5.1.1, cause I wasn't able to reproduce at all with other versions.

**Comment by lrineau** [comment]
TODO: try to reproduce the bug with g++-5.1, and to understand the context of the warning. Maybe a "static assert" in the method that triggers the warning would give more context...

**Comment by sgiraudot** [comment]
Is this issue still in the TODO list?

**Comment by efifogel** [comment]
Yes.
It's occurs only with g++5.1 (or higher).
It's about a compiler warning (related to deprecated code) that seemingly
should not be generated.
It looks like a compiler bug. If so, it should be reported.
If not, then the generic code that causes the warning is compiled in, while
it shouldn't. This might be caused by semantic changes in C++11. In this
case we need to fix the code so that it is not (compiled in).


   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by maxGimeno** [comment]
I don't see this warning in the testsuite anymore. Should we close this issue ?

**Comment by lrineau** [comment]
Was fixed by #649.
